engine/table/src/main/java/io/deephaven/engine/table/impl/perf/QueryPerformanceRecorder.java

```diff
@@ -100,128 +51,184 @@ public static void resetInstance() {
+     * Construct a QueryPerformanceRecorder for a top-level query.
+     *
+     * @param description the query description
+     * @param nuggetFactory the nugget factory
+     * @return a new QueryPerformanceRecorder
      */
-    public static void installUpdateGraphLockInstrumentation() {
-        UpdateGraphLock.installInstrumentation(new UpdateGraphLock.Instrumentation() {
-
-            @Override
-            public void recordAction(@NotNull String description, @NotNull Runnable action) {
-                QueryPerformanceRecorder.withNugget(description, action);
-            }
-
-            @Override
-            public void recordActionInterruptibly(@NotNull String description,
-                    @NotNull ThrowingRunnable<InterruptedException> action)
-                    throws InterruptedException {
-                QueryPerformanceRecorder.withNuggetThrowing(description, action);
-            }
-        });
+    static QueryPerformanceRecorder newQuery(
+            @NotNull final String description,
+            @NotNull final QueryPerformanceNugget.Factory nuggetFactory) {
+        return new QueryPerformanceRecorderImpl(description, null, nuggetFactory);
     }
 
     /**
-     * Record a single-threaded operation's allocations as "pool" allocated memory attributable to the current thread.
+     * Construct a QueryPerformanceRecorder for a sub-level query.
      *
-     * @param operation The operation to record allocation for
-     * @return The result of the operation.
+     * @param description the query description
+     * @param nuggetFactory the nugget factory
+     * @return a new QueryPerformanceRecorder
      */
-    public static <RESULT_TYPE> RESULT_TYPE recordPoolAllocation(@NotNull final Supplier<RESULT_TYPE> operation) {
-        final long startThreadAllocatedBytes = ThreadProfiler.DEFAULT.getCurrentThreadAllocatedBytes();
-        try {
-            return operation.get();
-        } finally {
-            final long endThreadAllocatedBytes = ThreadProfiler.DEFAULT.getCurrentThreadAllocatedBytes();
-            final MutableLong poolAllocatedBytesForCurrentThread = poolAllocatedBytes.get();
-            poolAllocatedBytesForCurrentThread.setValue(plus(poolAllocatedBytesForCurrentThread.longValue(),
-                    minus(endThreadAllocatedBytes, startThreadAllocatedBytes)));
-        }
+    static QueryPerformanceRecorder newSubQuery(
+            @NotNull final String description,
+            @Nullable final QueryPerformanceRecorder parent,
+            @NotNull final QueryPerformanceNugget.Factory nuggetFactory) {
+        return new QueryPerformanceRecorderImpl(description, parent, nuggetFactory);
     }
 
     /**
-     * Get the total bytes of pool-allocated memory attributed to this thread via
-     * {@link #recordPoolAllocation(Supplier)}.
+     * Starts a query.
+     * <p>
+     * It is an error to start a query more than once or while another query is running on this thread.
```

Comment:
A query is running if it has been {@link #startQuery() started} or {@link #resumed() resumed} without a subsequent {@link #endQuery() end}, {@link #suspendQuery() suspend}, or {@link #abortQuery() abort}).